### DISCUSSION TITLE
wasm-ld : error : --shared-memory is disallow

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#16825] [Q&A]
Hi,

my project must be compiled with pthread.
I tried to compile it with some static library (GDAL 2.0.0) that I built by emconfigure and emmake.
and I got the following error:
`wasm-ld : error : --shared-memory is disallowed by gt_jpeg_copy.o because it was not compiled with 'atomics' or 'bulk-memory' features.`

so I had set  the CFLAGS with -pthread or USE_PTHREADS  flag  before executing the emconfigure on GDAL , and I got the following error:
`checking whether we are cross compiling... (on node you may need: --experimental-wasm-threads --experimental-wasm-bulk-memory and also use a recent version)`

I will be grateful for any help to solve that issue.
Thanks

**Comment by tlively** [comment]
Is there any more to that second error message? If not, can you set `EMCC_VERBOSE=1` in the environment to get more information? Ultimately the problem is that `-pthread` is not being applied when compiling gt_jpeg_copy, and we have to figure out why.

**Comment by aharondavid** [comment]
there are some errors in the config.log . here is the config.log: 
[config.log](https://github.com/emscripten-core/emscripten/files/8569475/config.log)
 I turned on the EMCC_VERBOSE but I don't think that it add some additional info to the log file.

here is a section from the log that seem to me suspicious , but I do not know what it's mean
```
configure:3541: ./conftest
requested a shared WebAssembly.Memory but the returned buffer is not a SharedArrayBuffer, indicating that while the browser has SharedArrayBuffer it does not have WebAssembly threads support - you may need to set a flag
/mnt/c/engine/linux/GDAL/gdal-2.0.0/conftest:166
      throw ex;
      ^
```

**Comment by aharondavid** [comment]
eventually, I solved it by declared an environment: USER_DEFS=" -pthread" before emmake make.
Thanks.

**Reply by mightimatti** [reply]
Could you elaborate what you did specifically? I am experiencing the exact same issue while attempting to include `LibRaw` in my WASM project.  My understanding is that the `-pthread` flag needs to be added to the em++ calls, but I don't understand where I can set this flag. 
Adding `CFLAGS='-pthreads -s USE_PTHREADS=1' ` to the ./configure call failed so far

**Comment by aharondavid** [comment]
Hi,
here is my full bash file:
I hope this helps.
```
#!/bin/sh

cd ../emsdk
source ./emsdk_env.sh
export PATH=/mnt/c/engine/linux/emsdk:/mnt/c/engine/linux/emsdk/upstream/emscripten:/mnt/c/engine/linux/emsdk/node/14.18.2_64bit/bin:$PATH




export GDAL_BASE_DIR=/mnt/c/engine/linux/GDAL


cd gdal-3.1.2


export CFLAGS=" -I/mnt/c/engine/proj_dev/PROJ/src/  -I/mnt/c/Projects/TerraExplorer_v8_0_0_Dev/SLSource/Common/libxml/include -fexceptions -frtti -O3 "
export CXXFLAGS=" ${CFLAGS} -std=c++11"


emconfigure  ./configure  \
--with-proj=yes \
 --with-xml2=yes \
--with-ogr \
--without-sqlite3 \
--enable-shared=no \
--with-png=no \
--with-jpeg=no \
--disable-driver-mrf  \
--with-openjpeg=no  \
--without-jpeg12 \
--without-lerc \
--enable-debug=no


emmake make clean
export USER_DEFS=" -pthread"
emmake make -j8 

```

**Comment by mightimatti** [comment]
@aharondavid thank you. I'll test this when I'm back at my development machine.
